### PR TITLE
Reflection: task 1 and task 2

### PR DESCRIPTION
**Pull request by serhiiyasenev**

### CODE REVIEW

File: lesson6-Reflection/Task1/Container.cs [hunk 1]
```diff
@@ -1,28 +1,100 @@
...
 namespace Task1
 {
     public class Container
     {
+        private readonly Dictionary<Type, Type> _mappedTypes = new Dictionary<Type, Type>();
+
         public void AddAssembly(Assembly assembly)
         {
-            throw new NotImplementedException();
+            var mappedTypes = assembly.GetTypes().Where(t =>
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task1/Container.cs]: Extracting this logic to separate method may improve readability.

File: lesson6-Reflection/Task1/Container.cs [hunk 2]
```diff
@@ -1,28 +1,100 @@
...
+            if (!_mappedTypes.Any())
+                throw new Exception("No entity has been added yet");
+            return (T)Resolve(typeof(T));
+        }
+
+        private object Resolve(Type type)
+        {
+            Type resolvedType;
+            try
+            {
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task1/Container.cs]: Catching and throwing exception is resource-consuming operation, I'd recommend instead to throw exception if the type was not found in dictionary:
```suggestion
            if(!_mappedTypes.ContainsKey(type))
                throw new Exception(...);
```

File: lesson6-Reflection/Task1/Container.cs [hunk 3]
```diff
@@ -1,28 +1,100 @@
...
+                resolvedObject = ctor.Invoke(ctorParameters.Select(p => Resolve(p.ParameterType)).ToArray());
+            }
+            else
+            {
+                // If constructor hasn't parameter, create an instance of object
+                resolvedObject = Activator.CreateInstance(resolvedType); 
+            }
+            foreach (var prop in props)
+            {
+                prop.SetValue(resolvedObject,Resolve(prop.PropertyType));
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task1/Container.cs]: Missing space.

File: lesson6-Reflection/Task1/Container.cs [hunk 4]
```diff
@@ -1,28 +1,100 @@
...
+            }
+            catch(Exception resolveException)
+            {
+                throw new Exception("Unable to find type : " + type, resolveException);
+            }
+            // Inject marked properties
+            var props = resolvedType.GetProperties().Where(
+                prop => Attribute.IsDefined(prop, typeof(ImportAttribute)));
+            // Inject marked fields
+            var fields = resolvedType.GetFields().Where(
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task1/Container.cs]: In the task I see that there should be two possible options of injecting the dependency: through constructor and through property. So you shouldn't allow injecting directly through fields.

File: lesson6-Reflection/Task1/Container.cs [hunk 5]
```diff
@@ -1,28 +1,100 @@
...
+                // Inject marked classes
+                var ctor = resolvedType.GetConstructors().First();
+                var ctorParameters = ctor.GetParameters();
+                // Iterate through parameters and add each parameter
+                resolvedObject = ctor.Invoke(ctorParameters.Select(p => Resolve(p.ParameterType)).ToArray());
+            }
+            else
+            {
+                // If constructor hasn't parameter, create an instance of object
+                resolvedObject = Activator.CreateInstance(resolvedType);
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task1/Container.cs]: In the task it's stated that there may be only one option of injecting the dependency per single class. Therefore please move setting properties in this block.

File: lesson6-Reflection/Task1/Container.cs [hunk 6]
```diff
@@ -1,28 +1,100 @@
...
+            else
+            {
+                // If constructor hasn't parameter, create an instance of object
+                resolvedObject = Activator.CreateInstance(resolvedType); 
+            }
+            foreach (var prop in props)
+            {
+                prop.SetValue(resolvedObject,Resolve(prop.PropertyType));
+            }
+            foreach (var field in fields)
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task1/Container.cs]: This part is not needed as you shouldn't directly set fields.

File: lesson6-Reflection/Task1/Container.cs [hunk 7]
```diff
@@ -1,28 +1,100 @@
...
+        private readonly Dictionary<Type, Type> _mappedTypes = new Dictionary<Type, Type>();
+
         public void AddAssembly(Assembly assembly)
         {
-            throw new NotImplementedException();
+            var mappedTypes = assembly.GetTypes().Where(t =>
+                Attribute.IsDefined(t, typeof(ExportAttribute)) ||
+                Attribute.IsDefined(t, typeof(ImportConstructorAttribute)) ||
+                t.GetProperties().Any(p => Attribute.IsDefined(p, typeof(ImportAttribute))) ||
+                t.GetFields().Any(f => Attribute.IsDefined(f, typeof(ImportAttribute))));
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task1/Container.cs]: Part with `GetFields` is redundant.

File: lesson6-Reflection/Task1/Container.cs [hunk 8]
```diff
@@ -1,28 +1,100 @@
...
+
+            foreach (var type in mappedTypes)
+                if (type.GetCustomAttributes(typeof(ExportAttribute)).FirstOrDefault() == null)
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task1/Container.cs]:
```suggestion
		{
			var contract = (type.GetCustomAttributes(typeof(ExportAttribute)).FirstOrDefault() as ExportAttribute)
				?.Contract;
			if (contract != null)
				AddType(type, contract);
			else
				AddType(type);
		}
```
Reason: reading attributes through reflection is resource consuming, this way we can reduce number of such operations.

File: lesson6-Reflection/Task1/Container.cs [hunk 9]
```diff
@@ -1,28 +1,100 @@
...
+                        (type.GetCustomAttributes(typeof(ExportAttribute)).First() as ExportAttribute)?.Contract);
+                else
+                    AddType(type);
         }
 
         public void AddType(Type type)
         {
-            throw new NotImplementedException();
+            // If Already added, update the injection type
+            if (_mappedTypes.ContainsKey(type))
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task1/Container.cs]: Looking at the complex tests, if the type is already added and we try to add it again, it should throw an exception. So you should not do this check, it's enough just to add the type to dictionary.

**Review by YNauholnykh — CHANGES_REQUESTED** [review]

File: lesson6-Reflection/Task1/Container.cs (hunk 1 above)
Comment (serhiiyasenev) [reply lesson6-Reflection/Task1/Container.cs]: done

File: lesson6-Reflection/Task1/Container.cs (hunk 2 above)
Comment (serhiiyasenev) [reply lesson6-Reflection/Task1/Container.cs]: done

File: lesson6-Reflection/Task1/Container.cs (hunk 4 above)
Comment (serhiiyasenev) [reply lesson6-Reflection/Task1/Container.cs]: done

File: lesson6-Reflection/Task1/Container.cs (hunk 5 above)
Comment (serhiiyasenev) [reply lesson6-Reflection/Task1/Container.cs]: done

File: lesson6-Reflection/Task1/Container.cs (hunk 6 above)
Comment (serhiiyasenev) [reply lesson6-Reflection/Task1/Container.cs]: done

File: lesson6-Reflection/Task1/Container.cs (hunk 9 above)
Comment (serhiiyasenev) [reply lesson6-Reflection/Task1/Container.cs]: done

File: lesson6-Reflection/Task2/CustomArgumentNullException.cs [hunk 10]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Task2
+{
+    public class CustomException : Exception
+    {
+        public CustomException(string message) : base(message)
```
Comment (YNauholnykh) [inline lesson6-Reflection/Task2/CustomArgumentNullException.cs]: `MemberNotFoundException` sounds better for me. Also please don't forget to update the file name accordingly.

**Review by YNauholnykh — APPROVED** [review]
Approved with one minor comment.

File: lesson6-Reflection/Task2/CustomArgumentNullException.cs (hunk 10 above)
Comment (serhiiyasenev) [reply lesson6-Reflection/Task2/CustomArgumentNullException.cs]: done

**Pull request merged**